src/validation.cpp

```diff
@@ -582,6 +582,11 @@ class MemPoolAccept
         /** Total virtual size of all transactions being replaced. */
         size_t m_conflicting_size{0};
 
+        /** If we're doing package validation (i.e. m_package_feerates=true), the "effective"
+         * package feerate of this transaction, which may include fee and vsize of its ancestors
+         * and/or descendants. */
+        CFeeRate m_package_feerate{0};
```

Comment:
I'm wondering how well-defined this value is. Am I right in thinking that the calculation here is "sum(fees_of_package_transactions)/sum(size_of_package_transactions)"?

What happens in this case: we have 3 transactions A -> B, C, where C is a child of both A and B.

A has a low feerate and cannot make it in to the mempool by itself.  (A+B) has a high enough feerate for acceptance, but we receive for validation (A+B+C) instead.  B cannot make it in by itself, since it would be missing a parent, and we never try A+B, right?  So the package feerate for C would reflect the fees and feerate of B (which could be higher than C's own feerate, I think?).

Do we require that C pay a higher feerate than any of its parents when doing package acceptance?  If not, then I'm not sure this value will always make sense, ie the package feerate of a transaction with no children ought never be greater than its own feerate.

Reply:
Does this comment equally apply to package submission in master?

Reply:
I think so, but I was about to write a test using the example above, except where C is 0-fee, and see if it makes it in...

Reply:
Yes, master does not currently handle this case because we're currently doing A, B, C individually, so we'd get 1 "too low fee" and 2 "missing inputs," and then reject all of them. When I first tried to implement for 1-child-with-parents, I had incorrectly assumed this was okay and the code made it in before I realized - my apologies for that. The intention is definitely to fix it.

Requiring C pay a higher feerate than its parents is ok in the child-with-parents scenario but not generic ancestor package scenario (imagine low fee grandparent, lower fee parent, high fee child). I think we've discussed "for each tx in topological order, calculate its in-package ancestor set that hasn't made it in yet, and submit that as a package" - I have drafted this [here](https://github.com/glozow/bitcoin/commit/ef4e4936ab692d87dc0d5d55ac61e13940565947).

Btw this test case is implemented [here](https://github.com/glozow/bitcoin/commit/6659757cd547e4c3f6b62ef80cbc8cdda4b301b9) (it should fail on master as you expect).

Reply:
Can confirm the patch fails as expected: the entire triangle package makes it in the mempool

Reply:
Re: "I'm wondering how well-defined this value is." Last push adds the wtxids of which transactions are included in the feerate, to both `MempoolAcceptResult` and the testmempoolaccept/submitpackage RPCs.

Re: the problem where we don't always end up with the right transactions, might be better for another PR. I've opened #26711 to do this.

Reply:
@sdaftuar comfortable with resolving for this PR?